### PR TITLE
Address performance regression introduced by #90218

### PR DESCRIPTION
**Pull request by JakobDegen**
As part of the changes in #90218 , the `adt_drop_tys` and friends code stopped recursing through the query system, meaning that intermediate computations did not get cached. This change adds the recursions back in without re-introducing any of the old issues.

On local benchmarks this fixes the 5% regressions in #90504 ; the wg-grammar regressions didn't seem to move too much. I may take some time later to look into those.

Not sure who to request for review here, so will leave it up to whoever gets it.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
(Feel free to review as well, but I can try to help if you don't have a chance)

File: compiler/rustc_ty_utils/src/needs_drop.rs [hunk 1]
```diff
@@ -203,22 +210,46 @@ fn drop_tys_helper<'tcx>(
...
+                            } else {
+                                tcx.adt_drop_tys(adt_id.did)?
+                            }
+                            .into_iter()
+                            .map(|t| t.subst(tcx, subst)),
+                        );
+                    }
+                    _ => vec.push(subty),
+                };
+                Ok(vec)
```
Comment (the8472) [inline compiler/rustc_ty_utils/src/needs_drop.rs:250]: Imo a `for` loop would be more readable here.

File: compiler/rustc_ty_utils/src/needs_drop.rs [hunk 2]
```diff
@@ -186,11 +192,12 @@ fn drop_tys_helper<'tcx>(
...
     param_env: rustc_middle::ty::ParamEnv<'tcx>,
     adt_has_dtor: impl Fn(&ty::AdtDef) -> Option<DtorType>,
+    only_significant: bool,
 ) -> impl Iterator<Item = NeedsDropResult<Ty<'tcx>>> {
     let adt_components = move |adt_def: &ty::AdtDef, substs: SubstsRef<'tcx>| {
-        if adt_def.is_manually_drop() {
+        let mut subreqs = if adt_def.is_manually_drop() {
             debug!("drop_tys_helper: `{:?}` is manually drop", adt_def);
-            return Ok(Vec::new().into_iter());
+            Vec::new().into_iter()
```
Comment (the8472) [inline compiler/rustc_ty_utils/src/needs_drop.rs:200]: All those `into_iter` could be hoisted out of the if/else blocks

**Comment by JakobDegen** [comment]
Refactored more generally and extracted some of the uglyness out into a function, this both avoids the double allocation and makes things more readable

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

but I think this looks good to me -- presuming perf is still good after refactoring (as expected), r=me

**Comment by JakobDegen** [comment]
@Mark-Simulacrum perf seems to look good, but I believe I don't have perms to r= you

**Comment by JakobDegen** [comment]
Is this a genuine failure or spurious? I'm not terribly great at reading Bors logs

**Comment by the8472** [comment]
Segfault during bootstrap, CI has been having a few of those lately (#90812)

@bors retry

**Pull request merged**